samples/src/main/java/com/example/retrofit/PolymorphicConverter.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (C) 2017 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.example.retrofit;
+
+import java.io.IOException;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Type;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import okhttp3.RequestBody;
+import okio.Buffer;
+import retrofit2.Call;
+import retrofit2.Converter;
+import retrofit2.Retrofit;
+import retrofit2.converter.moshi.MoshiConverterFactory;
+import retrofit2.http.Body;
+import retrofit2.http.POST;
+
+/**
+ * A converter for polymorphic request bodies. This may be useful for libraries and places where the
+ * runtime types are unknown.
+ * If the runtime types are known, consider using separate overloaded methods on the service
+ * interface instead.
+ *
+ * For each service method, Retrofit caches converters for request bodies.
+ * The following requires loading from another cache for each invocation of the service method.
+ */
+final class PolymorphicConverter {
+  public static final class PolymorphicRequestBodyConverter<T>
+      implements Converter<T, RequestBody> {
+    public static final Factory FACTORY = new Factory() {
+      @Override public Converter<?, RequestBody> requestBodyConverter(Type type,
+          Annotation[] parameterAnnotations, Annotation[] methodAnnotations, Retrofit retrofit) {
+        return new PolymorphicRequestBodyConverter<>(this, parameterAnnotations, methodAnnotations,
```

Comment:
could have done something around `if (getRawType(type) != Animal.class) return null` to discourage using this everywhere, but that's just on CallAdapter.Factory.